torchdrivesim/behavior/iai.py

```diff
@@ -30,19 +32,84 @@ def iai_initialize(location, agent_count, center=(0, 0), traffic_light_state_his
-    response = invertedai.api.drive(
-        location=location, agent_states=agent_states, agent_attributes=agent_attributes,
-        recurrent_states=recurrent_states, traffic_lights_states=traffic_lights_states
+    try:
+        INITIALIZE_FOV = 120
+        conditional_agent_attributes = []
+        conditional_agent_states = []
+        conditional_recurrent_states = []
+        outside_agent_states = []
+        outside_agent_attributes = []
+        outside_recurrent_states = []
+
+        for i in range(len(agent_states)):
+            agent_state = agent_states[i]
+            dist = math.dist(center, (agent_state.center.x, agent_state.center.y))
+            if dist < INITIALIZE_FOV:
+                conditional_agent_states.append(agent_state)
+                conditional_agent_attributes.append(agent_attributes[i])
+                conditional_recurrent_states.append(recurrent_states[i])
+            else:
+                outside_agent_states.append(agent_state)
+                outside_agent_attributes.append(agent_attributes[i])
+                outside_recurrent_states.append(recurrent_states[i])
+
+        agent_count -= len(conditional_agent_states)
+        if agent_count > 0:
+            try:
+                seed = random.randint(1, 10000)
+                response = invertedai.api.initialize(
+                    location=location,
+                    agent_attributes=conditional_agent_attributes,
+                    states_history=[conditional_agent_states],
+                    agent_count=agent_count,
+                    location_of_interest=center,
+                    traffic_light_state_history=traffic_light_state_history,
+                    random_seed = seed
+                )
+                agent_attribute_list = response.agent_attributes + outside_agent_attributes
+                agent_state_list = response.agent_states + outside_agent_states
+                recurrent_state_list = response.recurrent_states + outside_recurrent_states
+            except Exception:
+                agent_attribute_list = agent_attributes
+                agent_state_list = agent_states
+                recurrent_state_list = recurrent_states
+        else:
+            agent_attribute_list = agent_attributes
+            agent_state_list = agent_states
+            recurrent_state_list = recurrent_states
+
+    except invertedai.error.InvalidRequestError:
+        raise InitializationFailedError()
+
+    agent_attributes = torch.stack(
+        [torch.tensor(at.tolist()[:3]) for at in agent_attribute_list], dim=-2
     )
     agent_states = torch.stack(
-        [torch.tensor(st.tolist()) for st in response.agent_states], dim=-2
+        [torch.tensor(st.tolist()) for st in agent_state_list], dim=-2
     )
+    return agent_attributes, agent_states, recurrent_state_list
+
+
+def iai_drive(location: str, agent_states: Tensor, agent_attributes: Tensor, recurrent_states: List, traffic_lights_states: Dict = None):
```

Comment:
Would you mind fixing the type signature while you're editing this function? It should specify the list element type for recurrent_state, make traffic_lights_states Optional, and specify return type.